### ISSUE TITLE
Errors in WSGI library aren't bubbled up to log

### ISSUE THREAD

**Issue by acedrew** [VOLTTRON/volttron#1387]
If there is an error retrieving data from an RPC endpoint from the web subsystem, no error is logged but the end http client receives a 500 response with this message: "Internal Server Error"
This error should make it to the logging system, I'm assuming this message is originating somewhere in the py4ws library, as I'm not seeing it anywhere in the volttron codebase.

**Comment by craig8** [comment]
Isn't this the trying to turn the None into a tuple issue that we saw?

**Comment by acedrew** [comment]
No, this is the fact that when we had that issue, the WSGI errors were getting dumped to STDOUT, without any appearance in the log that an error had occured. (I'm not necessarily recommending that we pipe all of the WSGI output to the log, but we should log that an WSGI error happened)

**Comment by craig8** [comment]
You know that file that is output the wsgi error log that we looked at and had the file object <w>...if that wasn't there you would see everything in stdout if I remember.  Of course I have been wrong before.

**Comment by acedrew** [comment]
I don't need it in standard out, I need some indication that something is wrong in the logger being used by volttron, I'll look at how the wsgi server is started by volttron, and if we can watch for stderr or stdout there to bubble up a log.

**Comment by acedrew** [comment]
Oh wow, this is actually a much bigger issue than I had hoped. The underlying gevent.pywsgi can be initialized with a 'error_log' kwarg that is a file like object for errors to be written to, however, none of the error handling in gevent.pywsgi actually writes to it.

**Comment by acedrew** [comment]
I've posted a message to the gevent list to see what their recommendation is.

**Comment by craig8** [comment]
Can you reference the list/issue you sent that to?

**Comment by acedrew** [comment]
@craig8 http://www.gevent.org/community.html
It hasn't been approved when last I checked

**Comment by acedrew** [comment]
> Hello all, I attempted capture WSGI errors by passing a StringIO object to the 'error_log' kwarg on initialization of the WSGIServer class, however there doesn't seem to be any data written to the object. 

Right. `wsgi.errors` is provided for the *application* (wsgi callable) to write to. According to the specification, it may or may not interact with the server's default error logging[1]. 

> In digging through the code, I see several comments in the error handling about not using the wsgi.errors property despite it's presence. I also see where traceback.exc() is being used rather than writing to the wsgi.errors object. 

Mostly right. Client errors (that happen before the application is invoked due to bad HTTP input, etc) are directed to `wsgi.errors`. Errors raised by the application are not. 

gevent has the established pattern of directing unhandled errors through the event loop's `handle_error` method---which in reality defaults to the hub's handle_error method---and pywsgi follows this pattern. 


> Any ideas on how I can capture wsgi errors into my logging handler in my application? 


The simplest and most local thing to do would be to subclass pywsgi.WSGIHandler  and override `_log_error`. (That's not a documented method, though, so it may change in future versions.) 

  class MyHandler(WSGIHandler): 
    def _log_error(self, t, v, tb): 
       self.server.error_log.write(format_exception(t, v, tb)) 

  WSGIServer(":8080", app, handler_class=MyHandler).serve_forever() 

The currently officially supported way to deal with all uncaught errors within gevent is to set a different error handler for the loop: 

  orig_error_handler = get_hub().loop.error_handler 

  class MyErrorHandler(object): 
     def handle_error(self, context, t, v, tb): 
        if want_to_handle_context(context): 
            logging.getLogger('my.logger').exception(...) 
        else: 
            orig_error_handler(context, t, v, tb) 

  get_hub().loop.error_handler = MyErrorHandler() 

Alternatively (and probably what I would recommend for "production" apps), use a WSGI middleware wrapped around your app to catch unhandled exceptions so that they simply don't bubble up to gevent at all. It could do something as simple as logging (repoze.errorlog does this[2]), or it could be as complex as the one provided with paste[3]; I'm sure there are others. Frameworks like Pyramid do something like that by default, IIRC, so you can also look into that. 

Jason 

[1] https://www.python.org/dev/peps/pep-3333/#environ-variables 
[2] http://repozeerrorlog.readthedocs.io/en/latest/ 
[2] https://bitbucket.org/ianb/paste/src/0e5a48796ab969d874c6b772c5c33561ac2d1b0d/paste/exceptions/errormiddleware.py?at=default&fileviewer=file-view-default

**Comment by acedrew** [comment]
OK, I'm working on this, it looks like we pass an error handler to the websockets server, but not the the generic WSGI server, this should be a simple fix, though may break in the future because that interface is not part of the specification.

**Comment by craig8** [comment]
Yeah great.  as I read that I was like ... hey we have middleware already with the __call__ stuff that should be easy :)
